File: pt5/1140.stone-game-ii.java
import java.util.*;

class Solution {
    private int[] suffix;
    private int[][] memo;
    private int n;

    public int stoneGameII(int[] piles) {
        n = piles.length;
        suffix = new int[n + 1];
        for (int i = n - 1; i >= 0; i--) {
            suffix[i] = suffix[i + 1] + piles[i];
        }

        memo = new int[n][n + 1];
        for (int[] row : memo) {
            Arrays.fill(row, -1);
        }
        return bestFrom(0, 1);
    }

    private int bestFrom(int index, int m) {
        if (index >= n) {
            return 0;
        }
        if (2 * m >= n - index) {
            return suffix[index];
        }
        if (memo[index][m] != -1) {
            return memo[index][m];
        }

        int best = 0;
        for (int x = 1; x <= 2 * m; x++) {
            int opponent = bestFrom(index + x, Math.max(m, x));
            best = Math.max(best, suffix[index] - opponent);
        }
        memo[index][m] = best;
        return best;
    }
}

File: pt5/1140.stone-game-ii.py
from __future__ import annotations

from functools import lru_cache
from typing import List


class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        n = len(piles)
        suffix = [0] * (n + 1)
        for i in range(n - 1, -1, -1):
            suffix[i] = suffix[i + 1] + piles[i]

        @lru_cache(None)
        def best_from(i: int, m: int) -> int:
            if i >= n:
                return 0
            if 2 * m >= n - i:
                return suffix[i]

            best = 0
            for x in range(1, 2 * m + 1):
                opponent = best_from(i + x, max(m, x))
                best = max(best, suffix[i] - opponent)
            return best

        return best_from(0, 1)

